force-app/main/default/classes/DMLOperations.cls

```diff
@@ -146,8 +160,65 @@ public with sharing class DMLOperations {
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
         // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        // Create an Opportunity with the given name
+        // Add the Opportunity to the list
+
+        // Case 1: Found existing Account with the given accountName -> assign Account Id to accountRecord variable
+        // Case 2: Create a new Account record with the given accountName
+        List<Account> accountList = [SELECT Id FROM Account WHERE Name = :accountName Limit 1];
+        Account accountRecord = new Account();
+
+        if (!accountList.isEmpty()) {
+            accountRecord.Id = accountList.get(0).Id;
+        } else {
+            accountRecord.Name = accountName;
+        }
+        upsert accountRecord;
+
+        // Case 1a: Some Opportunity records (existingOpportunities) already exist with the given names -> Assign accountRecord.Id to those opportunities
+        // Case 1b: Some Opportunity records (newOppRecords) do not exist with the given names -> Create new Opportunity records with the given names -> Assign accountRecord.Id to new opportunities 
+        // Case 2: For all the given OppNames, create new opportunities -> Assign accountRecord.Id to those opportunities
+        List<Opportunity> existingOpportunities = new List<Opportunity>();
+        if (!oppNames.isEmpty()) {
+            existingOpportunities = [SELECT Id, Name, AccountId FROM Opportunity WHERE Name IN :oppNames];
+        }
+        List<String> existingOppNames = new List<String>();
+        List<String> newOppNames = new List<String>();
+        List<Opportunity> newOppRecords = new List<Opportunity>();
+        if (!existingOpportunities.isEmpty()) {
+            for (Opportunity currentOppRecord : existingOpportunities) {
+                if (currentOppRecord.AccountId != accountRecord.Id) {
+                    currentOppRecord.AccountId = accountRecord.Id;
+                }
+                existingOppNames.add(currentOppRecord.Name);
+            }     
+            upsert existingOpportunities;
+            for (String currentOppName1 : oppNames) {
+                if (!existingOppNames.contains(currentOppName1)) {
+                    newOppNames.add(currentOppName1);
+
+                }
+            }
```

Comment:
Very good job avoiding DML within the loop and checking for the ones that needed to be created.